Nodal Patch Recovery Aux

hugary1995 opened the pull request:
Elemental patch recovery that fits nodal values for gradient and material properties

The main work is done in NodalPatchRecovery.C which derives from AuxKernel
NodalPatchRecovery.C
  -- generates multi-index for a complete set of monomials
  -- assembles the patch stiffness matrix and load vector
  -- performs the least square fitting
  -- uses the fitted coefficients to calculate the nodal value
  -- leaves ComputeValue() abstract

Sample usage of the AuxKernel is illustrated in StressRecovery.C and NodalGradientAux.C which derives from NodalPatchRecovery and implement the ComputeValue() method.

A flag was added in AuxKernel.C to let it know whether the derived AuxKernel is using NodalPatchRecovery. If the flag is set to be true, then it won't issue error when retrieving material properties, gradients, etc.

Corresponding regression tests have been added.

refs #11880

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/213571/) on fdc6f1f wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11881/site)

This comment will be updated on new commits.

YaqiWang commented:
We possibly should also do a check to make sure that no patched nodal variable depends on another patched nodal variable if we do the element loop once for all aux kernels. If we keep this way, i.e. each aux kernel has its own element loop, the dependency will be fine. Sorry for my comments, this is a great capability.

hugary1995 committed:
nodal patch recovery auxkernel

Elemental patch recovery that fits nodal values for gradient and material properties

refs #11880

Co-authored-by: Wen Jiang <jiangwen84@gmail.com>

bwspenc commented on modules/tensor_mechanics/src/auxkernels/StressRecovery.C:
```diff
@@ -0,0 +1,35 @@
+/****************************************************************/
```
Some might disagree with me, but I don't really like the idea of having two separate classes for getting the stress (one for element and one for nodal values).  Could we make it so that NodalPatchRecovery can also run on element AuxVariables, and then make RankTwoScalarAux derive from it, and have it work in both cases?

bwspenc commented on modules/tensor_mechanics/test/tests/stress_recovery/patch/patch.i:
```diff
@@ -0,0 +1,150 @@
+[Executioner]
+  type = Transient
+  solve_type = NEWTON
+  petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
```
Don't use superlu unless you really need it (because it's not necessarily compiled in).

permcody replied:
If you do, make sure you add super lu to the test spec so that we won't attempt to run this test on a machine that doesn't have it.

bwspenc commented on framework/src/auxkernels/NodalPatchRecovery.C:
```diff
@@ -0,0 +1,208 @@
+  MeshBase & meshhelper = _mesh.getMesh();
+  meshhelper.allow_renumbering(false);
+  for (const auto & elem :
+       as_range(meshhelper.semilocal_elements_begin(), meshhelper.semilocal_elements_end()))
```
I take it that this fixed the parallel problems? Are you essentially ghosting out a second layer of elements?

hugary1995 replied:
You are right I am fixing the parallel issue here. However, this is not ghosting out a second layer.
It is a bit complicated to explain without drawing any graph. Here I am ghosting out all semilocal elements -- which is different from simply ghosting one layer out in a very subtle way.

I am also trying to get distributed mesh to work using RelationshipManager.

permcody replied:
I recommend you stay away from semilocal elements. It's not quite what you think it is, trust me. Try searching for it in the tickets! 

The right way is with a relationship manager as you've already said. There's already existing objects that will do arbitrary numbers of layers or point neighbors and then you are good to go.

dschwen replied:
@hugary1995 are you trying to get _point neighbors_ here? Because that would require at least *2* additional ghosting layers with quad elements, and possibly more with tets etc! You also need to be super careful with periodic boundary conditions.

permcody replied:
This problem is solved: See how I do it for the GrainTracker:
https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/postprocessors/GrainTracker.C#L55

https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/relationshipmanagers/GrainTrackerHaloRM.C

Now, your use case will be a little more complex because you actually need solution data off from the ghosted element correct? For that, you'll need to create an AlgebraicRelationshipManager and we'll have to figure out exactly what we'll need to do to get solution information. This is probably also a solved problem but if not, I'm willing to help you reach that solution. Please stop by and chat when you are ready to work on this. @roystgnr will be helpful in well in telling us 1) if a ghosting functor exists that already gets solution information from n layers deep 2) Is that built into the "DefaultCoupling" ghosting functor already? Sorry, too lazy to look at the moment.

roystgnr replied:
It depends on how you define "layers".  DefaultCoupling and PointNeighborCoupling both have set_n_levels(int) methods; if you set that to e.g. 2 then with DefaultCoupling you get side-neighbors-of-side-neighbors and with PointNeighborCoupling you get point-neighbors-of-point-neighbors.

hugary1995 replied:
@permcody I'll have time to work on this next week. What would be the best time for you to meet?

bwspenc reviewed (changes requested):
A few things to address, but this is getting there.

hugary1995 commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,96 @@
+  /**
+   * generate all combinations of n choose k
+   */
+  virtual std::vector<std::vector<unsigned>> nchoosek(unsigned N, unsigned K);
```
I think N choose K is a quite basic and fundamental algorithm. Do you think it can go into MooseUtil?

permcody replied:
Yes, but I wouldn't want this implementation. If we add this to MOOSE it should be rewritten to be a generator instead. Building and returning all results in vectors can be _very_ space expensive.

dschwen replied:
This is already in libMesh as `Utility::binomial(n , k)`!

dschwen replied:
https://github.com/libMesh/libmesh/blob/master/include/utils/utility.h#L219-L237

hugary1995 replied:
@dschwen Utility::binomial(n, k) computes the number of combinations of n choose k. What I am trying to do here is to return a listing of all combinations -- maybe I should use a more descriptive name here.

dschwen replied:
Ha, right, I should have at least checked the return types...

permcody commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,96 @@
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
```
Wrong header - make sure you update all of them.

hugary1995 committed:
address comments and regold

hugary1995 commented:
@bwspenc All touches in AuxKernel are now removed. I also derived RankTwoAux and RankTwoScalarAux from NodalPatchRecovery so they work with either nodal variables or elemental variables. The default setting in TensorMechanicsAction hasn't been changed yet -- it is still running on constant monomial.

@permcody I rolled back to manual ghosting and added TODOs in the code so that people are aware of the relationshipmanager issue. Tests are still only run on replicated mesh until we switch to relationshipmanager.

hugary1995 commented on framework/src/auxkernels/NodalPatchRecovery.C:
```diff
@@ -0,0 +1,208 @@
+validParams<NodalPatchRecovery>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addRequiredParam<unsigned>("order", "polynomial order");
```
This is the only comment I did not address. Here the order means polynomial order of the least squares fitting. I think it is different from the variable order, i.e. this kernel is always run on a first lagrange variable but the polynomial order is user-defined.

bwspenc commented:
A few things to address based on our discussion:

1. Change 'ploynomial_order' -> 'patch_polynomial_order', and make its default be the variable's order.
2. Add a couple of sentences in the documentation of RankTwoAux and RankTwoScalarAux to explain what happens when you have a nodal AuxVariable.
3. Test on higher order variables.

bwspenc commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,130 @@
+public:
+  NodalPatchRecovery(const InputParameters & parameters);
+  virtual ~NodalPatchRecovery(){};
+  /// Override from Auxkernel to suppress warnings
```
All of these methods need documentation that follows the form shown under 'Documentation' in.
http://mooseframework.org/wiki/CodeStandards/
Use the '/**' style, and describe all parameters and returns using `@param` and `@return`.

bwspenc commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,130 @@
+
+protected:
+  /// generate all combinations of n choose k
+  virtual std::vector<std::vector<unsigned>> nchoosek(unsigned N, unsigned K);
```
nchoosek -> nChooseK to follow MOOSE naming conventions.

bwspenc replied:
Per MOOSE code standards, change all instances of 'unsigned' -> 'unsigned int'

bwspenc commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,130 @@
+  virtual void compute() override;
+
+  /// a boolean indicating whether a patch polynomial order is specified by the user
+  const bool _order_provided;
```
Since `_order_provided` isn't used anywhere but in the constructor, I'd get rid of it.

bwspenc commented on framework/src/auxkernels/NodalPatchRecovery.C:
```diff
@@ -0,0 +1,224 @@
+NodalPatchRecovery::NodalPatchRecovery(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _order_provided(isParamValid("patch_polynomial_order")),
+    _order(_order_provided ? getParam<unsigned>("patch_polynomial_order") : (unsigned)_var.order()),
```
_order(isParamValid("patch_polynomial_order") ? getParam<unsigned>("patch_polynomial_order") : (unsigned)_var.order()),

bwspenc commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,130 @@
+  const bool _order_provided;
+
+  /// polynomial order, default is variable order
+  const unsigned _order;
```
`_order` -> `_patch_polynomial_order` (to avoid confusion with variable order).

bwspenc commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,130 @@
+  const unsigned _order;
+
+  /// mesh dimension
+  const unsigned _dim;
```
Consider getting rid of `_dim` as a member variable as well since it's also only used in the constructor.

bwspenc commented on framework/src/auxkernels/NodalPatchRecovery.C:
```diff
@@ -0,0 +1,224 @@
+validParams<NodalPatchRecovery>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addParam<unsigned>("patch_polynomial_order", "polynomial order");
```
Add a better description of this:
Polynomial order used in least squares fitting of material property over the local patch of elements connected to a given node

bwspenc commented on framework/src/auxkernels/NodalPatchRecovery.C:
```diff
@@ -0,0 +1,224 @@
+  // TODO remove the manual ghosting once relationship manager is working correctly
+  // no need to ghost if this aux is elemental
+  if (isNodal())
+  {
```
Would it make sense to have a check to make sure that the user didn't specify a patch polynomial order lower than the variable order?
```
   if (isParamValid("patch_polynomial_order"))
    mooseWarning("Specified 'patch_polynomial_order' is lower than the AuxVariable's order");
```

hugary1995 replied:
I think it is okay at least in some cases to specify a patch polynomial order higher than the variable order as we discussed earlier. I do have a check to make sure we have enough patch sample points to do the least squares fitting:

```
// before we go, check if we have enough sample points for solving the least square fitting
  if (_q_point.size() * elem_ids.size() < _multi_index.size())
    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
               "polynomial order or using a higher-order quadrature scheme.");
```

bwspenc commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,130 @@
+
+  /**
+   * generate a complete multi index table for given dimension and order
+   * i.e. given dim = 2, order = 2, generated table will have the following content
```
Either here or in the documentation for _multi_index, explain what the entries mean.

bwspenc commented on framework/include/auxkernels/NodalPatchRecovery.h:
```diff
@@ -0,0 +1,130 @@
+ * Zienkiewicz, O. C. and Zhu, J. Z. (1992), The superconvergent patch recovery and a posteriori
+ * error estimates. Part 1: The recovery technique. Int. J. Numer. Meth. Engng., 33: 1331-1364.
+ * doi:10.1002/nme.1620330702
+ */
```
Add this to the description:
If used with elemental (monomial) AuxVariables, this class falls back on the standard techniques in AuxKernel.C to compute either constant or higher-order monomials, depending on the order of the AuxVariable.

bwspenc commented on modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md:
```diff
@@ -37,13 +37,16 @@ AuxVariable is used as the argument for the `variable` input parameter in the `R
-higher-order elemental variables, these AuxVariables are output by libMesh as nodal variables for
-visualization purposes. Using higher order monomial variables in this way can produce smoother
-visualizations of results for a properly converged simulation.
+the AuxVariable for the entire element, which is computed by taking a local $L_2$ projection. Using an
```
Is the original 'volume-weighted average' statement wrong? I think it is more generally understood.

hugary1995 replied:
I haven't heard of it before... Anyways I have rolled it back to volume-weighted average.

bwspenc commented on modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md:
```diff
@@ -37,13 +37,16 @@ AuxVariable is used as the argument for the `variable` input parameter in the `R
+
+Using an AuxVariable with `family = LAGRANGE` will give a smooth nodal field of the material property,
+constructed using nodal patch recovery. `patch_polynomial_order` is set to equal the order of the AuxVariable
+by default.
```
I think I would re-arrange the above text a little to convey the fact that there are really 3 main options: constant monomial, higher order monomial, and patch recovery for nodal variables. Emphasize that for the best (smoothest, most accurate) results, use the last option, but there is some additional computational cost.

bwspenc commented on framework/src/auxkernels/NodalPatchRecovery.C:
```diff
@@ -0,0 +1,224 @@
+void
+NodalPatchRecovery::compute()
+{
+  if (!isNodal())
```
A comment here indicating what we are doing with this switch would be helpful (just something simple like "Fall back on standard procedure for element variables", and "Use Zienkiewicz Zu patch recovery for nodal variables").

bwspenc reviewed (changes requested):
This is looking really close.  Some nit-picky requests.

bwspenc commented:
@permcody I think this is getting really close.  I'm ready to approve this as soon as he addresses my issues. I'd like to make sure we get your buy-in on this architecture before we merge this, so if you could look this over, I'd appreciate it.  We can discuss this tomorrow.

hugary1995 committed:
address more comments

bwspenc commented on modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoAux.md:
```diff
@@ -36,14 +36,27 @@ AuxVariable is used as the argument for the `variable` input parameter in the `R
-visualizations of results for a properly converged simulation.
+Results will have different quality based on the AuxVariable:
+
+  * Elemental Constant Monomial
```
Try something like this:
```
* _Element Constant Monomial_ Using an AuxVariable ...
* _Elemental Higher-order Monomial_ blah blah blah...

hugary1995 committed:
address more more comments

hugary1995 committed:
modify documentation
